File: src/proxy/raw_websocket.rs
use rand::Rng;
use tokio::io::{AsyncBufReadExt, AsyncReadExt, AsyncWriteExt, BufReader};
use tokio::net::TcpStream;
use tokio_native_tls::TlsConnector;

use crate::proxy::crypto::xor_mask;

type TlsHalf = tokio::io::ReadHalf<tokio_native_tls::TlsStream<TcpStream>>;

fn tls_connector() -> TlsConnector {
    let mut builder = native_tls::TlsConnector::builder();
    builder.danger_accept_invalid_certs(true);
    TlsConnector::from(builder.build().unwrap())
}

#[derive(Debug)]
pub struct WsHandshakeError {
    pub status_code: u16,
    pub status_line: String,
    pub location: Option<String>,
}

impl WsHandshakeError {
    pub fn is_redirect(&self) -> bool {
        matches!(self.status_code, 301 | 302 | 303 | 307 | 308)
    }
}

impl std::fmt::Display for WsHandshakeError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "HTTP {}: {}", self.status_code, self.status_line)
    }
}

impl std::error::Error for WsHandshakeError {}

pub struct RawWebSocket {
    reader: BufReader<TlsHalf>,
    writer: tokio::io::WriteHalf<tokio_native_tls::TlsStream<TcpStream>>,
    closed: bool,
}

impl RawWebSocket {
    pub async fn connect(
        host: &str,
        domain: &str,
        timeout: std::time::Duration,
        path: &str,
    ) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
        let stream = tokio::time::timeout(timeout, TcpStream::connect((host, 443))).await??;
        stream.set_nodelay(true)?;

        let connector = tls_connector();
        let tls_stream =
            tokio::time::timeout(timeout, connector.connect(domain, stream)).await??;

        let (reader, writer) = tokio::io::split(tls_stream);
        let mut reader = BufReader::new(reader);

        let ws_key = {
            let mut bytes = [0u8; 16];
            rand::rng().fill_bytes(&mut bytes);
            base64::Engine::encode(&base64::engine::general_purpose::STANDARD, bytes)
        };

        let req = format!(
            "GET {path} HTTP/1.1\r\n\
             Host: {domain}\r\n\
             Upgrade: websocket\r\n\
             Connection: Upgrade\r\n\
             Sec-WebSocket-Key: {ws_key}\r\n\
             Sec-WebSocket-Version: 13\r\n\
             Sec-WebSocket-Protocol: binary\r\n\
             \r\n"
        );

        let mut writer = writer;
        writer.write_all(req.as_bytes()).await?;
        writer.flush().await?;

        let mut response_lines = Vec::new();
        loop {
            let mut line = String::new();
            let n = tokio::time::timeout(timeout, reader.read_line(&mut line)).await??;
            if n == 0 || line.trim().is_empty() {
                break;
            }
            response_lines.push(line.trim().to_string());
        }

        if response_lines.is_empty() {
            return Err("empty response".into());
        }

        let first_line = &response_lines[0];
        let parts: Vec<&str> = first_line.splitn(3, ' ').collect();
        let status_code: u16 = parts.get(1).and_then(|s| s.parse().ok()).unwrap_or(0);

        if status_code == 101 {
            return Ok(Self {
                reader,
                writer,
                closed: false,
            });
        }

        let location = response_lines[1..]
            .iter()
            .find_map(|line| {
                let mut parts = line.splitn(2, ':');
                let key = parts.next()?.trim().to_lowercase();
                if key == "location" {
                    Some(parts.next()?.trim().to_string())
                } else {
                    None
                }
            });

        Err(WsHandshakeError {
            status_code,
            status_line: first_line.clone(),
            location,
        }
        .into())
    }

    pub async fn send(&mut self, data: &[u8]) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        if self.closed {
            return Err("WebSocket closed".into());
        }
        let frame = build_frame(0x2, data, true);
        self.writer.write_all(&frame).await?;
        self.writer.flush().await?;
        Ok(())
    }

    pub async fn send_batch(
        &mut self,
        parts: &[Vec<u8>],
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        if self.closed {
            return Err("WebSocket closed".into());
        }
        for part in parts {
            let frame = build_frame(0x2, part, true);
            self.writer.write_all(&frame).await?;
        }
        self.writer.flush().await?;
        Ok(())
    }

    pub async fn recv(&mut self) -> Result<Option<Vec<u8>>, Box<dyn std::error::Error + Send + Sync>> {
        loop {
            let (opcode, payload) = read_frame(&mut self.reader).await?;

            match opcode {
                0x8 => {
                    self.closed = true;
                    let close_frame = build_frame(0x8, &payload[..payload.len().min(2)], true);
                    let _ = self.writer.write_all(&close_frame).await;
                    let _ = self.writer.flush().await;
                    return Ok(None);
                }
                0x9 => {
                    let pong = build_frame(0xA, &payload, true);
                    let _ = self.writer.write_all(&pong).await;
                    let _ = self.writer.flush().await;
                    continue;
                }
                0xA => continue,
                0x1 | 0x2 => return Ok(Some(payload)),
                _ => continue,
            }
        }
    }

    pub async fn close(&mut self) {
        if self.closed {
            return;
        }
        self.closed = true;
        let frame = build_frame(0x8, &[], true);
        let _ = self.writer.write_all(&frame).await;
        let _ = self.writer.flush().await;
        let _ = self.writer.shutdown().await;
    }

    pub fn is_closed(&self) -> bool {
        self.closed
    }
}

pub fn build_frame(opcode: u8, data: &[u8], mask: bool) -> Vec<u8> {
    let len = data.len();
    let mut frame = Vec::with_capacity(10 + len);
    frame.push(0x80 | opcode);

    if mask {
        let mask_key: [u8; 4] = rand::random();
        if len < 126 {
            frame.push(0x80 | len as u8);
        } else if len < 65536 {
            frame.push(0x80 | 126);
            frame.extend_from_slice(&(len as u16).to_be_bytes());
        } else {
            frame.push(0x80 | 127);
            frame.extend_from_slice(&(len as u64).to_be_bytes());
        }
        frame.extend_from_slice(&mask_key);
        frame.extend_from_slice(&xor_mask(data, &mask_key));
    } else {
        if len < 126 {
            frame.push(len as u8);
        } else if len < 65536 {
            frame.push(126);
            frame.extend_from_slice(&(len as u16).to_be_bytes());
        } else {
            frame.push(127);
            frame.extend_from_slice(&(len as u64).to_be_bytes());
        }
        frame.extend_from_slice(data);
    }
    frame
}

pub async fn read_frame(
    reader: &mut BufReader<TlsHalf>,
) -> Result<(u8, Vec<u8>), Box<dyn std::error::Error + Send + Sync>> {
    let mut hdr = [0u8; 2];
    reader.read_exact(&mut hdr).await?;
    let opcode = hdr[0] & 0x0F;
    let mut length = (hdr[1] & 0x7F) as u64;

    if length == 126 {
        let mut buf = [0u8; 2];
        reader.read_exact(&mut buf).await?;
        length = u16::from_be_bytes(buf) as u64;
    } else if length == 127 {
        let mut buf = [0u8; 8];
        reader.read_exact(&mut buf).await?;
        length = u64::from_be_bytes(buf);
    }

    let payload = if hdr[1] & 0x80 != 0 {
        let mut mask_key = [0u8; 4];
        reader.read_exact(&mut mask_key).await?;
        let mut data = vec![0u8; length as usize];
        reader.read_exact(&mut data).await?;
        xor_mask(&data, &mask_key)
    } else {
        let mut data = vec![0u8; length as usize];
        reader.read_exact(&mut data).await?;
        data
    };

    Ok((opcode, payload))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_build_frame() {
        let data = b"hello";
        let frame = build_frame(0x2, data, true);
        assert!(frame.len() > data.len());
        assert_eq!(frame[0] & 0x0F, 0x2);
        assert!(frame[1] & 0x80 != 0);
    }

    #[test]
    fn test_build_frame_unmasked() {
        let data = b"hello";
        let frame = build_frame(0x2, data, false);
        assert_eq!(frame[0] & 0x0F, 0x2);
        assert!(frame[1] & 0x80 == 0);
        assert_eq!(&frame[2..], data);
    }
}

File: proxy/raw_websocket.py
import os
import ssl
import base64
import struct
import asyncio
import socket as _socket

from typing import List, Optional, Tuple
from .config import proxy_config


_st_BB = struct.Struct('>BB')
_st_BBH = struct.Struct('>BBH')
_st_BBQ = struct.Struct('>BBQ')
_st_BB4s = struct.Struct('>BB4s')
_st_BBH4s = struct.Struct('>BBH4s')
_st_BBQ4s = struct.Struct('>BBQ4s')
_st_H = struct.Struct('>H')
_st_Q = struct.Struct('>Q')

_ssl_ctx = ssl.create_default_context()
_ssl_ctx.check_hostname = False
_ssl_ctx.verify_mode = ssl.CERT_NONE


class WsHandshakeError(Exception):
    def __init__(self, status_code: int, status_line: str,
                 headers: Optional[dict] = None, location: Optional[str] = None):
        self.status_code = status_code
        self.status_line = status_line
        self.headers = headers or {}
        self.location = location
        super().__init__(f"HTTP {status_code}: {status_line}")

    @property
    def is_redirect(self) -> bool:
        return self.status_code in (301, 302, 303, 307, 308)


# Pre-allocated mask key — WS mask only needs to be unpredictable on the wire;
# the payload is already AES-CTR encrypted so a fixed mask is safe and saves
# one os.urandom() syscall per frame on the hot path.
_MASK_KEY = os.urandom(4)
_MASK_KEY_INT = int.from_bytes(_MASK_KEY, 'big')


try:
    import tg_ws_proxy_rs
    _rust_xor_mask = tg_ws_proxy_rs.xor_mask
except ImportError:
    _rust_xor_mask = None


def _xor_mask(data: bytes, mask: bytes) -> bytes:
    """XOR data with a 4-byte repeating mask key (WS RFC 6455).

    Uses compiled Rust module if available, otherwise falls back to fast struct-based Python XOR.
    """
    if _rust_xor_mask is not None:
        return _rust_xor_mask(data, mask)

    n = len(data)
    if not n:
        return data
    # Unpack mask as one uint32
    m32 = struct.unpack('>I', mask)[0]
    out = bytearray(n)
    # Process 4 bytes at a time
    full, rem = divmod(n, 4)
    off = 0
    for _ in range(full):
        v = struct.unpack_from('>I', data, off)[0] ^ m32
        struct.pack_into('>I', out, off, v)
        off += 4
    # Handle remainder (0–3 bytes)
    for i in range(rem):
        out[off + i] = data[off + i] ^ mask[i]
    return bytes(out)


def set_sock_opts(transport, buffer_size):
    sock = transport.get_extra_info('socket')
    if sock is None:
        return
    
    try:
        sock.setsockopt(_socket.IPPROTO_TCP, _socket.TCP_NODELAY, 1)
    except (OSError, AttributeError):
        pass
    
    try:
        sock.setsockopt(_socket.SOL_SOCKET, _socket.SO_RCVBUF, buffer_size)
        sock.setsockopt(_socket.SOL_SOCKET, _socket.SO_SNDBUF, buffer_size)
    except OSError:
        pass


class RawWebSocket:
    __slots__ = ('reader', 'writer', '_closed')

    OP_BINARY = 0x2
    OP_CLOSE = 0x8
    OP_PING = 0x9
    OP_PONG = 0xA

    def __init__(self, reader: asyncio.StreamReader,
                 writer: asyncio.StreamWriter):
        self.reader = reader
        self.writer = writer
        self._closed = False

    @staticmethod
    async def connect(host: str, domain: str, timeout: float = 10.0,
                      path: str = '/apiws') -> 'RawWebSocket':
        reader, writer = await asyncio.wait_for(
            asyncio.open_connection(host, 443, ssl=_ssl_ctx,
                                    server_hostname=domain),
            timeout=min(timeout, 10))
        
        set_sock_opts(writer.transport, proxy_config.buffer_size)

        ws_key = base64.b64encode(os.urandom(16)).decode()

        req = (
            f'GET {path} HTTP/1.1\r\n'
            f'Host: {domain}\r\n'
            f'Upgrade: websocket\r\n'
            f'Connection: Upgrade\r\n'
            f'Sec-WebSocket-Key: {ws_key}\r\n'
            f'Sec-WebSocket-Version: 13\r\n'
            f'Sec-WebSocket-Protocol: binary\r\n'
            f'\r\n'
        )
        writer.write(req.encode())
        await writer.drain()

        response_lines: list[str] = []
        try:
            while True:
                line = await asyncio.wait_for(reader.readline(),
                                              timeout=timeout)
                if line in (b'\r\n', b'\n', b''):
                    break
                response_lines.append(
                    line.decode('utf-8', errors='replace').strip())
        except asyncio.TimeoutError:
            writer.close()
            raise

        if not response_lines:
            writer.close()
            raise WsHandshakeError(0, 'empty response')

        first_line = response_lines[0]
        parts = first_line.split(' ', 2)
        try:
            status_code = int(parts[1]) if len(parts) >= 2 else 0
        except ValueError:
            status_code = 0

        if status_code == 101:
            return RawWebSocket(reader, writer)

        headers: dict[str, str] = {}
        for hl in response_lines[1:]:
            if ':' in hl:
                k, v = hl.split(':', 1)
                headers[k.strip().lower()] = v.strip()

        writer.close()
        raise WsHandshakeError(status_code, first_line, headers,
                                location=headers.get('location'))

    async def send(self, data: bytes) -> None:
        """Write one binary WS frame; drain is managed by the bridge loop."""
        if self._closed:
            raise ConnectionError("WebSocket closed")
        self.writer.write(self._build_frame(self.OP_BINARY, data, mask=True))

    async def send_batch(self, parts: List[bytes]) -> None:
        """Write multiple binary WS frames in one go without intermediate drains."""
        if self._closed:
            raise ConnectionError("WebSocket closed")
        # Concatenate all frames into a single write call to minimise syscalls
        buf = bytearray()
        for part in parts:
            buf += self._build_frame(self.OP_BINARY, part, mask=True)
        self.writer.write(bytes(buf))

    async def recv(self) -> Optional[bytes]:
        while not self._closed:
            opcode, payload = await self._read_frame()

            if opcode == self.OP_CLOSE:
                self._closed = True
                try:
                    self.writer.write(self._build_frame(
                        self.OP_CLOSE,
                        payload[:2] if payload else b'', mask=True))
                    await self.writer.drain()
                except Exception:
                    pass
                return None

            if opcode == self.OP_PING:
                try:
                    self.writer.write(
                        self._build_frame(self.OP_PONG, payload, mask=True))
                    await self.writer.drain()
                except Exception:
                    pass
                continue

            if opcode == self.OP_PONG:
                continue

            if opcode in (0x1, 0x2):
                return payload
            continue
        return None

    async def close(self):
        if self._closed:
            return
        self._closed = True
        try:
            self.writer.write(
                self._build_frame(self.OP_CLOSE, b'', mask=True))
            await self.writer.drain()
        except Exception:
            pass
        try:
            self.writer.close()
            await self.writer.wait_closed()
        except Exception:
            pass

    @staticmethod
    def _build_frame(opcode: int, data: bytes, mask: bool = False) -> bytes:
        length = len(data)
        fb = 0x80 | opcode
        if not mask:
            if length < 126:
                return _st_BB.pack(fb, length) + data
            if length < 65536:
                return _st_BBH.pack(fb, 126, length) + data
            return _st_BBQ.pack(fb, 127, length) + data
        # Reuse the module-level mask key — payload is already AES-CTR encrypted
        mask_key = _MASK_KEY
        masked = _xor_mask(data, mask_key)
        if length < 126:
            return _st_BB4s.pack(fb, 0x80 | length, mask_key) + masked
        if length < 65536:
            return _st_BBH4s.pack(fb, 0x80 | 126, length, mask_key) + masked
        return _st_BBQ4s.pack(fb, 0x80 | 127, length, mask_key) + masked

    async def _read_frame(self) -> Tuple[int, bytes]:
        hdr = await self.reader.readexactly(2)
        opcode = hdr[0] & 0x0F
        length = hdr[1] & 0x7F
        if length == 126:
            length = _st_H.unpack(await self.reader.readexactly(2))[0]
        elif length == 127:
            length = _st_Q.unpack(await self.reader.readexactly(8))[0]
        if hdr[1] & 0x80:
            mask_key = await self.reader.readexactly(4)
            payload = await self.reader.readexactly(length)
            return opcode, _xor_mask(payload, mask_key)
        payload = await self.reader.readexactly(length)
        return opcode, payload